### ISSUE TITLE
Different results in content encoding in munit with amp O1 or amp O0

### ISSUE THREAD

**Issue by theponpon** [NVlabs/imaginaire#11]
Hi there,

First of all thank you for this amazing library.
It really helps people like me to bootstrap in the amazing world of generative networks!

That said I have noticed a strange behaviour when running the training on munit/afhq_doc2cat:
- when you run in amp O1 optimization level, the content reconstruction error diverges (error above 3.5 all the time)
- however running exactly the same settings but with amp O0, the same content reconstruction error converges to values as small as 0.6
See attached content_recon.png
![content_recon](https://user-images.githubusercontent.com/984998/95688640-9628d800-0bd9-11eb-9f1e-37208e9bd6cc.png)

I wonder if this is the expected behaviour.

Additional information:
- the same behaviour happens when using torch.cuda.amp instead of apex
- the same behaviour happens when using my dataset which is not about dogs and cats ;)
- the style encoding reconstruction error does NOT seem to suffer from the same issue
See attached style_recon.png
![style_recon](https://user-images.githubusercontent.com/984998/95688642-96c16e80-0bd9-11eb-90b5-60253075beb8.png)

Configuration:
- ubuntu 18.04
- 2 V100 GPUs
- nvidia driver 450.66
- pytorch 1.6
- cuda 10.2.89
- cudnn 8.04.30

Looking forward to reading from you,

Pierre

**Comment by mingyuliutw** [comment]
Could you try to use NVIDIA docker we provide on the installation markdown page? We have tested MUNIT with torch amp and it works for us. But we are using nvidia pytorch docker on ubuntu 16.04. Another difference is that we are using 8 V100 GPUs. We haven't tested it with 2 V100 GPUs.

**Comment by xunhuang1995** [comment]
I remember I had similar issues before. Can you try to replace `torch.abs` with `nn.L1Loss` in loss computation (https://github.com/NVlabs/imaginaire/blob/master/imaginaire/trainers/munit.py#L164) and see if that makes a difference?

**Comment by theponpon** [comment]
@xunhuang1995 : that seems to do the trick! Thanks a lot!
Note 1: I would never suspect this part of the code! Curious to get an explanation of why this happens to avoid the trap in the future.
Note 2: great job with MUNIT!

**Comment by xunhuang1995** [comment]
Thanks for catching this. I have fixed it in the newest commit. Though I'm still unsure what's the exact reason behind it :(

**Comment by theponpon** [comment]
Hi there,
I am afraid the fix committed yesterday has a little typo:
line 156:
instead of
         self.gen_losses['style_recon_b'] = \
it should be 
        self.gen_losses['content_recon_b'] = \

As is today, the code crashes on line 150 as self.gen_losses['content_recon_b'] is not defined

**Comment by xunhuang1995** [comment]
Thanks for noticing that. Just fixed it.
